### PR TITLE
CurveBs|tools-v2|rebuild check-consistency,list may-broken-vol, check-chunkserver

### PR DESCRIPTION
**Pull request by jolly-sy**
the content of this pr is about rebuilding three command :check-consistency,list may-broken-vol, check-chunkserver

check-consistency:
image

list may broken vol:
image

check- chunkserver
image

### CODE REVIEW

**Comment by jolly-sy** [comment]
cicheck

**Comment by Cyber-SiKu** [comment]
All output should be printed using tableNew, please do not print by yourself, this output is not friendly.
If you need to print multiple tables, please consider splitting them into multiple commands and pay attention to the uniform style of the tables.
----
所有的输出都应该使用tableNew 打印，请不要自行打印，这样输出不友好。
如果需要打印多个表格，请考虑拆分成多个命令，注意表格风格的统一。

File: tools-v2/internal/error/error.go [hunk 1]
```diff
@@ -357,7 +357,7 @@ var (
 		return NewInternalCmdError(36, "get time of day fail, the error is: %s")
 	}
 	ErrBsGetFileInfo = func() *CmdError {
-		return NewInternalCmdError(37, "get file info fail, the error is: %s")
+		return NewInternalCmdError(37, "%s")
```
Comment (Cyber-SiKu) [inline tools-v2/internal/error/error.go:360]: why delete this？

File: tools-v2/internal/proto/curvebs/topology/statuscode.proto [hunk 2]
```diff
@@ -38,4 +38,42 @@ enum TopoStatusCode {
...
+    CrcFail = 3;                    //Crc校验错误
+    InvalidReqParam = 4;           //请求参数错误
+    NoEnoughSpace = 5;            //空间不够
+    CopysetNotExist = 6;         //copyset不存在
+    ChunkNotExist = 7;          //chunk或其快照文件不存在
+    UnknowFailure = 8;          //未知错误
+    OverLoad = 9;               //服务端过载
+    Backward = 10;             // 请求的版本落后当前chunk的版本
+    ChunkExist = 11;           //chunk已存在
+    EpochOld = 12;             // 请求中的epoch太老了
```
Comment (Cyber-SiKu) [inline tools-v2/internal/proto/curvebs/topology/statuscode.proto:71]: English and alignment

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+	        or
+	        curve bs check chunkserver --chunkserverAddr=127.0.0.1:8200`
+)
+
+const (
+	STATE_TRANSFERRING = "TRANSFERRING"
+	STATE_CANDIDATE    = "CANDIDATE"
+	kCpsHealthy        = 0  // chunkserver上所有copyset健康
+	kNotHealthy        = -1 // chunkserver上有copyset不健康
+	kNotOnline         = -2 // chunkserver不在线
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go:30]: english

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+	}
+	unhealthyCnt, r := cksCmd.CountUnHealthyCopysets()
+	row := make(map[string]string)
+	total := strconv.FormatUint(uint64(len(cksCmd.total)), 10)
+	unCnt := strconv.FormatUint(uint64(unhealthyCnt), 10)
+	ratio := strconv.FormatFloat(r, 'f', 2, 64) + "%"
+	// fmt.Println("ratio is", ratio)
+	row["total copysets"] = total
+	row["unhealthy copysets"] = unCnt
+	row["unhealthy ratio"] = ratio
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go:169]: add in internal/utils/row.go

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	kParseError = -1
+	// peer数量小于预期
+	kPeersNoSufficient = -2
+	// 副本间的index差距太大
+	kLogIndexGapTooBig = -3
+	// 有副本在安装快照
+	kInstallingSnapshot = -4
+	// 少数副本不在线
+	kMinorityPeerNotOnline = -5
+	// 大多数副本不在线
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go:31]: english

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	if len(copysetInfos) == 0 {
+		fmt.Println("copysetInfo is empty")
+		return err1
+	}
+	lpid := copysetInfos[0].GetLogicalPoolId()
+	var cpsIds []uint32
+	for _, cpsInfo := range copysetInfos {
+		cpsIds = append(cpsIds, cpsInfo.GetCopysetId())
+	}
+	// fmt.Println("cpsIds is", cpsIds)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go:175]: rm

**Review by Cyber-SiKu — CHANGES_REQUESTED** [review]

**Comment by jolly-sy** [comment]
> 

这条命令会根据情况产生多种不一样的结果，表头无法统一，所以无法使用表格统一打印

**Comment by Cyber-SiKu** [comment]
> > 
> 
> 这条命令会根据情况产生多种不一样的结果，表头无法统一，所以无法使用表格统一打印

The table header is not necessarily formulated in init, you can add the header when you are sure.

**Comment by jolly-sy** [comment]
cicheck

File: proto/topology.proto [hunk 7]
```diff
@@ -334,11 +344,49 @@ message PhysicalPoolResponse {
...
+    optional string desc = 5;
+}
+
+message PoolsetResponse {
+    required sint32 statusCode = 1;
+    optional PoolsetInfo poolsetInfo = 2;
+}
+
+
+
```
Comment (Cyber-SiKu) [inline proto/topology.proto:379]: remove extra blank lines

File: proto/topology.proto [hunk 8]
```diff
@@ -334,11 +344,49 @@ message PhysicalPoolResponse {
...
+message ListPoolsetRequest{
+
+}
+
+message ListPoolsetResponse{
+    required sint32 statusCode = 1;
+    repeated PoolsetInfo poolsetInfos = 2;
+}
+
```
Comment (Cyber-SiKu) [inline proto/topology.proto:388]: remove extra blank lines

File: tools-v2/testTopo.json [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{
```
Comment (Cyber-SiKu) [inline tools-v2/testTopo.json:1]: move to tools-v2/pkg/config/

File: tools-v2/pkg/config/template.yaml [hunk 10]
```diff
@@ -23,3 +23,4 @@ curvebs:
   root:
     user: root
     password: root_password
+
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/template.yaml:26]: remove extra blank lines

File: tools-v2/pkg/config/bs.go [hunk 11]
```diff
@@ -33,26 +34,44 @@ import (
...
+	VIPER_CURVEBS_ETCDADDR         = "curvebs.etcdAddr"
+	CURVEBS_PATH                   = "path"
+	VIPER_CURVEBS_PATH             = "curvebs.path"
+	CURVEBS_USER                   = "user"
+	VIPER_CURVEBS_USER             = "curvebs.root.user"
+	CURVEBS_DEFAULT_USER           = "root"
+	CURVEBS_PASSWORD               = "password"
+	VIPER_CURVEBS_PASSWORD         = "curvebs.root.password"
+	CURVEBS_DEFAULT_PASSWORD       = "root_password"
+	CURVEBS_CLUSTERMAP             = "bsClusterMap"
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:51]: lowercase

File: tools-v2/pkg/config/bs.go [hunk 12]
```diff
@@ -33,26 +34,44 @@ import (
...
+	VIPER_CURVEBS_FILENAME         = "curvebs.filename"
+	CURVEBS_HASH_CHECK             = "hashcheck"
+	VIPER_CURVEBS_HASH_CHECK       = "curvebs.hashcheck"
+	CURVEBS_CHUNK_SIZE             = "chunkSize"
+	VIPER_CURVEBS_CHUNK_SIZE       = "curvebs.chunkSize"
+	CURVEBS_PORT                   = "port"
+	VIPER_CURVEBS_PORT             = "curvebs.port"
+	CURVEBS_CHUNKSERVER_ID         = "csId"
+	VIPER_CURVEBS_CHUNKSERVER_ID   = "curvebs.csId"
+	CURVEBS_CHUNKSERVER_ADDR       = "csAddr"
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:67]: ditto

File: tools-v2/pkg/config/bs.go [hunk 13]
```diff
@@ -33,26 +34,44 @@ import (
...
+	VIPER_CURVEBS_CHUNKSERVER_ADDR = "curvebs.csAddr"
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:68]: ditto

File: tools-v2/pkg/config/bs.go [hunk 14]
```diff
@@ -33,26 +34,44 @@ import (
...
+	VIPER_CURVEBS_DIR              = "curvebs.dir"
+	CURVEBS_FILENAME               = "filename"
+	VIPER_CURVEBS_FILENAME         = "curvebs.filename"
+	CURVEBS_HASH_CHECK             = "hashcheck"
+	VIPER_CURVEBS_HASH_CHECK       = "curvebs.hashcheck"
+	CURVEBS_CHUNK_SIZE             = "chunkSize"
+	VIPER_CURVEBS_CHUNK_SIZE       = "curvebs.chunkSize"
+	CURVEBS_PORT                   = "port"
+	VIPER_CURVEBS_PORT             = "curvebs.port"
+	CURVEBS_CHUNKSERVER_ID         = "csId"
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:65]: ditto

File: tools-v2/pkg/cli/curvecli.go [hunk 15]
```diff
@@ -47,7 +47,7 @@ func setupRootCommand(cmd *cobra.Command) {
 }
 
 func newCurveCommand() *cobra.Command {
-	cmd := &cobra.Command{
+	rootCmd := &cobra.Command{
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/curvecli.go:50]: why rename？

File: tools-v2/pkg/cli/command/curvefs/create/topology/zone.go [hunk 16]
```diff
@@ -166,7 +166,7 @@ func (tCmd *TopologyCommand) scanZones() *cmderror.CmdError {
 
 func (tCmd *TopologyCommand) removeZones() *cmderror.CmdError {
 	tCmd.deleteZoneRpc = &DeleteZoneRpc{}
-	tCmd.deleteZoneRpc.Info = basecmd.NewRpc(tCmd.addrs, tCmd.timeout, tCmd.retryTimes, "DeleteServer")
+	tCmd.deleteZoneRpc.Info = basecmd.NewRpc(tCmd.addrs, tCmd.timeout, tCmd.retryTimes, "DeleteZone")
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvefs/create/topology/zone.go:169]: good job！

File: tools-v2/pkg/cli/command/curvebs/status/etcd/etcd.go [hunk 17]
```diff
@@ -189,26 +189,26 @@ func (eCmd *EtcdCommand) ResultPlainOutput() error {
 }
 
 func NewStatusEtcdCommand() *EtcdCommand {
-	etcdCmd := &EtcdCommand{
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/status/etcd/etcd.go:192]: why rename？

File: tools-v2/pkg/cli/command/curvebs/query/file/fileinfo.go [hunk 18]
```diff
@@ -103,7 +103,7 @@ func (gCmd *GetFileInfoCommand) Init(cmd *cobra.Command, args []string) error {
 	// fmt.Println("user:", owner)
 	// fmt.Println("viper:", viper.GetString(config.VIPER_CURVEBS_USER))
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/query/file/fileinfo.go:104]: you can delete this

File: tools-v2/pkg/cli/command/curvebs/create/cluster/topo.go [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+	if err.TypeCode() != cmderror.CODE_SUCCESS {
+		return err
+	}
+	return cmderror.ErrSuccess()
+}
+
+func (ctCmd *ClusterTopoCmd) RunCommand(cmd *cobra.Command, args []string) error {
+	err := ctCmd.updateTopology()
+	if err.TypeCode() != cmderror.CODE_SUCCESS {
+		//return fmt.Errorf(err.Message)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/create/cluster/topo.go:254]: remove

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+func (ciRpc *GetChunkInfoRpc) NewRpcClient(cc grpc.ClientConnInterface) {
+	ciRpc.ChunkClient = chunk.NewChunkServiceClient(cc)
+}
+
+func (ciRpc *GetChunkInfoRpc) Stub_Func(ctx context.Context) (interface{}, error) {
+	return ciRpc.ChunkClient.GetChunkInfo(ctx, ciRpc.Request)
+}
+
+func (mdvCmd *MayDamVolCmd) ListChunkServersOnServers() *cmderror.CmdError {
+	// fmt.Println("ListChunkServersOnServers() in")
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:45]: rm

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+		if response.GetStatusCode() != 0 {
+			return cmderror.ErrListChunkServers(statuscode.TopoStatusCode(response.GetStatusCode()))
+		}
+
+		for _, chunkServer := range response.GetChunkServerInfos() {
+			if chunkServer.GetStatus() == *topology.ChunkServerStatus_RETIRED.Enum() {
+				mdvCmd.isRetired = true
+				continue
+			}
+			//fmt.Println(chunkServer)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:70]: rm

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+			mdvCmd.chunkServerInfos = append(mdvCmd.chunkServerInfos, chunkServer)
+		}
+	}
+	return cmderror.ErrSuccess()
+}
+
+func (mdvCmd *MayDamVolCmd) CheckIfChunkServerOnline(csAddr string) bool {
+	// fmt.Printf("CheckIfChunkServerOnline() in, csAddr is csAddr %s", csAddr)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:78]: rm

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	var s []string
+	s = append(s, csAddr)
+	mdvCmd.getChunkInfoRpc.Info = basecmd.NewRpc(s, mdvCmd.timeout, mdvCmd.retryTimes, "GetChunkInfo")
+	_, err := basecmd.GetRpcResponse(mdvCmd.getChunkInfoRpc.Info, mdvCmd.getChunkInfoRpc)
+	return err.TypeCode() == cmderror.CODE_SUCCESS
+}
+
+func (mdvCmd *MayDamVolCmd) FindOffLineChunkServers() *cmderror.CmdError {
+	// fmt.Println("FindOffLineChunkServers() in")
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:98]: rm

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+	if len(mdvCmd.offLineCsAddrs) == 0 {
+		// fmt.Println("len(mdvCmd.offLineCsAddrs) == 0")
+		return err.ToError()
+	}
+	// 获取所有离线server上的copysets
+	err = mdvCmd.GetCopySetsOnOffLineCs()
+	if err.TypeCode() != cmderror.CODE_SUCCESS {
+		return err.ToError()
+	}
+	//统计各个copysets出现的次数,2次及以上的视为损坏的copyset，记录进mdvCmd.damagedCps
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go:116]: English note

**Review by Cyber-SiKu — COMMENTED** [review]
delete tools-v2/pkg/cli/command/curvebs/list/topo

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+	fileNames         []string
+	isRetired         bool
+
+	rows []map[string]string
+}
+
+func NewMayDamVolCmd() *cobra.Command {
+	mdvCmd := &MayDamVolCmd{
+		FinalCurveCmd: basecmd.FinalCurveCmd{
+			Use:     "may-damaged-vols",
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go:60]: damaged-vols may be better？

File: tools-v2/internal/error/error.go (hunk 1 above)
Comment (jolly-sy) [reply tools-v2/internal/error/error.go]: > why delete this？

coding error, fix

File: tools-v2/internal/proto/curvebs/topology/statuscode.proto (hunk 2 above)
Comment (jolly-sy) [reply tools-v2/internal/proto/curvebs/topology/statuscode.proto:71]: > English and alignment

fix

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go (hunk 3 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go:30]: > english

fix

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go (hunk 4 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/check/chunkserver/chunkserver.go:169]: > add in internal/utils/row.go

fix

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go (hunk 5 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go:31]: > english

fix

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go (hunk 6 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/check/chunkserver/copyset.go:175]: > rm

fix

File: proto/topology.proto (hunk 7 above)
Comment (jolly-sy) [reply proto/topology.proto:379]: > remove extra blank lines

fix

File: proto/topology.proto (hunk 8 above)
Comment (jolly-sy) [reply proto/topology.proto:388]: > remove extra blank lines

fix

File: tools-v2/testTopo.json (hunk 9 above)
Comment (jolly-sy) [reply tools-v2/testTopo.json:1]: > move to tools-v2/pkg/config/

fix

File: tools-v2/pkg/config/bs.go (hunk 11 above)
Comment (jolly-sy) [reply tools-v2/pkg/config/bs.go:51]: > lowercase

all lowercase is ambiguous

File: tools-v2/pkg/config/bs.go (hunk 13 above)
Comment (jolly-sy) [reply tools-v2/pkg/config/bs.go:68]: > ditto

all lowercase is ambiguous

File: tools-v2/pkg/config/bs.go (hunk 14 above)
Comment (jolly-sy) [reply tools-v2/pkg/config/bs.go:65]: > ditto

all lowercase is ambiguous

File: tools-v2/pkg/cli/curvecli.go (hunk 15 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/curvecli.go:50]: > why rename？

for test , I will change it back

File: tools-v2/pkg/cli/command/curvebs/status/etcd/etcd.go (hunk 17 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/status/etcd/etcd.go:192]: > why rename？

coding error. fix

File: tools-v2/pkg/cli/command/curvebs/query/file/fileinfo.go (hunk 18 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/query/file/fileinfo.go:104]: > you can delete this

fix

File: tools-v2/pkg/cli/command/curvebs/create/cluster/topo.go (hunk 19 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/create/cluster/topo.go:254]: > remove

fix

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go (hunk 20 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:45]: > rm

fix

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go (hunk 21 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:70]: > rm

fix

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go (hunk 22 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:78]: > rm

fix

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go (hunk 23 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/chunkserver.go:98]: > rm

fix

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go (hunk 24 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go:116]: > English note

fix

File: tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go (hunk 25 above)
Comment (jolly-sy) [reply tools-v2/pkg/cli/command/curvebs/list/maydamagedvolumes/maydamagedvolumes.go:60]: > `damaged-vols` may be better？

no, original name is may-broken-vol, may can't be removed

**Comment by aspirer** [comment]
we should split these commands and commit them one by one.

**Pull request closed without merging**